File: mercury_v2_backend/app/schemas/canvas.py
from pydantic import BaseModel, ConfigDict, Field
from datetime import datetime
from typing import Optional
import uuid


# Canvas Project Schemas
class CanvasProjectCreate(BaseModel):
    """Schema for creating a canvas project."""
    name: str = Field(..., min_length=1, max_length=200)
    project_id: Optional[uuid.UUID] = None


class CanvasProjectUpdate(BaseModel):
    """Schema for updating canvas state."""
    canvas_state: Optional[dict] = None


class CanvasProjectResponse(BaseModel):
    """Schema for canvas project response."""
    id: uuid.UUID
    name: str
    project_id: Optional[uuid.UUID]
    canvas_state: Optional[dict]
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


# Canvas Layer Schemas
class CanvasLayerCreate(BaseModel):
    """Schema for creating a canvas layer."""
    layer_type: str = Field(..., min_length=1, max_length=50)
    layer_data: dict
    z_index: int
    is_visible: bool = True
    is_locked: bool = False
    opacity: float = Field(default=1.0, ge=0.0, le=1.0)


class CanvasLayerUpdate(BaseModel):
    """Schema for updating a canvas layer."""
    layer_data: Optional[dict] = None
    is_visible: Optional[bool] = None
    is_locked: Optional[bool] = None
    opacity: Optional[float] = Field(None, ge=0.0, le=1.0)


class CanvasLayerResponse(BaseModel):
    """Schema for canvas layer response."""
    id: uuid.UUID
    layer_type: str
    layer_data: dict
    z_index: int
    is_visible: bool
    is_locked: bool
    opacity: float
    created_at: datetime
    updated_at: datetime
    
    model_config = ConfigDict(from_attributes=True)


class CanvasProjectDetailResponse(BaseModel):
    """Schema for detailed canvas project response with layers."""
    id: uuid.UUID
    name: str
    project_id: Optional[uuid.UUID]
    canvas_state: Optional[dict]
    layers: list[CanvasLayerResponse]
    created_at: datetime
    updated_at: datetime


# Segmentation Schemas
class SegmentData(BaseModel):
    """Schema for a single segment."""
    id: str
    label: str
    mask_data: dict
    color: str


class SegmentationRequest(BaseModel):
    """Schema for segmentation request."""
    layer_id: uuid.UUID


class SegmentationResponse(BaseModel):
    """Schema for segmentation response."""
    segments: list[SegmentData]


# Generation Schemas
class GenerationParams(BaseModel):
    """Schema for generation parameters."""
    strength: Optional[float] = Field(None, ge=0.0, le=1.0)
    steps: Optional[int] = Field(None, ge=1, le=150)
    guidance_scale: Optional[float] = Field(None, ge=0.0, le=20.0)


class SketchToImageRequest(BaseModel):
    """Schema for sketch-to-image generation request."""
    layer_ids: list[uuid.UUID]
    prompt: str
    generation_params: Optional[GenerationParams] = None


class InpaintingRequest(BaseModel):
    """Schema for inpainting request."""
    layer_id: uuid.UUID
    mask_data: dict
    prompt: str
    generation_params: Optional[GenerationParams] = None


class TaskResponse(BaseModel):
    """Schema for async task response."""
    task_id: str
    status: str

File: mercury_v2_frontend/src/types/api/canvas.ts
export interface CanvasInstance {
  id: string;
  name: string;
  project_id?: string;
  canvas_state: {
    viewport: {
      x: number;
      y: number;
      zoom: number;
    };
  };
  thumbnail_url: string;
  created_at: string;
  updated_at: string;
  layers?: CanvasLayer[];
}

export type CanvasProject = CanvasInstance;

export type LayerType = 'sketch' | 'image' | 'text';

export type ImageSource = 'upload' | 'ai_generated' | 'chat_import';

export interface BaseLayer {
  id: string;
  layer_type: LayerType;
  z_index: number;
  is_visible: boolean;
  is_locked?: boolean;
  opacity?: number;
  name?: string;
  created_at?: string;
  updated_at?: string;
}

export interface SketchLayerData {
  paths: any[]; // Fabric.js path objects
  x?: number;
  y?: number;
  width?: number;
  height?: number;
  fabric_json?: any;
}

export interface SketchLayer extends BaseLayer {
  layer_type: 'sketch';
  layer_data: SketchLayerData;
}

export interface ImageLayerData {
  image_url: string;
  source: ImageSource;
  x: number;
  y: number;
  width: number;
  height: number;
  scale_x?: number;
  scale_y?: number;
  rotation?: number;
  
  // AI Generated
  parent_layer_id?: string;
  prompt?: string;
  generation_params?: {
    strength?: number;
    steps?: number;
    guidance_scale?: number;
    model?: string;
  };
  reference_layer_ids?: string[];
  
  // Chat Import
  chat_message_id?: string;
  
  fabric_json?: any;
}

export interface ImageLayer extends BaseLayer {
  layer_type: 'image';
  layer_data: ImageLayerData;
}

export interface TextLayerData {
  text: string;
  x: number;
  y: number;
  font_family?: string;
  font_size?: number;
  fill?: string;
  width?: number;
  height?: number;
  fabric_json?: any;
}

export interface TextLayer extends BaseLayer {
  layer_type: 'text';
  layer_data: TextLayerData;
}

export type CanvasLayer = SketchLayer | ImageLayer | TextLayer;

export interface Segment {
  id: string;
  label: string;
  mask_data: {
    paths: any[];
  };
  color: string;
}

export interface SegmentationResult {
  segments: Segment[];
}

export interface AsyncTaskResponse {
  task_id: string;
  status: 'processing' | 'completed' | 'failed';
  progress?: number;
  result?: any;
  error?: any;
}

export interface TaskResponse {
    task_id: string;
    status: 'PENDING' | 'PROGRESS' | 'SUCCESS' | 'FAILURE';
}

// Request DTOs
export interface CreateCanvasProjectRequest {
  name: string;
  project_id?: string;
}

export interface UpdateCanvasProjectRequest {
  name?: string;
  canvas_state?: {
    viewport: {
      x: number;
      y: number;
      zoom: number;
    };
  };
}

export interface CreateLayerRequest {
  layer_type: LayerType;
  layer_data: SketchLayerData | ImageLayerData | TextLayerData;
  z_index: number;
}

export interface UpdateLayerRequest {
  layer_data?: Partial<SketchLayerData | ImageLayerData | TextLayerData>;
  is_visible?: boolean;
  is_locked?: boolean;
  opacity?: number;
  z_index?: number;
}

export interface GenerateImageRequest {
  layer_ids: string[];
  prompt: string;
  generation_params?: {
    strength?: number;
    steps?: number;
    guidance_scale?: number;
    model?: string;
  };
}

export interface InpaintRequest {
  layer_id: string;
  mask_data: {
    paths: any[];
  };
  prompt: string;
  generation_params?: {
    strength?: number;
  };
}

export type ShapeType = 'rect' | 'circle' | 'polygon' | 'line';